### PR TITLE
Add CONTRIBUTING information for new issues

### PR DESCRIPTION
**Pull request by gmelikov**
I copied "Reporting a problem" section from the wiki's FAQ to issue template and added `Please review the guidelines for contributing to this repository.` info, i think it makes the issuing process easier for new people.

Example - https://github.com/gmelikov/zfs/issues/new

### CODE REVIEW

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+**What distribution you're using and the version.**
```
Comment (mmlb) [inline .github/ISSUE_TEMPLATE.md]: I would drop all the `.` they look weird when rendered.

File: .github/ISSUE_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+**What spl/zfs packages you're using and the version.**
```
Comment (mmlb) [inline .github/ISSUE_TEMPLATE.md]: I'd reword this and previous to be `you're` -> `are you`. You are reads like a directive, but you are asking a question.

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+### Reporting a problem
+
```
Comment (mmlb) [inline .github/CONTRIBUTING.md]: Paragraphs can be split at each sentence so that they are not huge one-liners.

**Comment by bunder2015** [comment]
"which" sounds better than "what"

**Comment by gmelikov** [comment]
Thanks for comments, fixed.

**Comment by gmelikov** [comment]
@behlendorf fully agree with you, i'll make a draft.

Waiting for any suggestions.

**Comment by gmelikov** [comment]
I added the structure and some information. My comments marked as italic.

Please, write any additions here, i'll add them.

**Comment by gmelikov** [comment]
I made some additions, you can see them [here](https://github.com/gmelikov/zfs/blob/master/.github/CONTRIBUTING.md).

I appreciate any addition about these moments:
* Pull Requests requirements
* Styleguides
* Debugging info (for example, how to profile performance issues)

I made minimal variant, which i can write by myself. I think we can merge it because of issue template, it'll probably help us with issues quality.

**Comment by tuxoko** [comment]
@gmelikov 
Can you add in template that log file should be marked with \`\`\`
Thanks.

**Comment by gmelikov** [comment]
@tuxoko done.

**Comment by mmlb** [comment]
It may make sense to have put the note about the ``` in as a html comment. The reporter will see the note and not need remember to remove it. We can also add any other messages that are only pertinent to the report at initial report time.

**Comment by gmelikov** [comment]
@mmlb your variant is well too, but i think if it will be a comment then reporter can easily skip it.
If the reporter has to delete the note, he'll 100% see it and will do as needed.

I'll appreciate someone's else opinion on this question, both variants are good for me.

**Comment by mmlb** [comment]
Right that's possible, but I don't think it's likely. Reporter not knowing/using ``` for logs/long input is not likely to be the common scenario. It's very likely that issue reporters will either do so already or only need the gentlest reminder, otoh making the common case always remove it can get to be annoying.

I've had those feelings while working in another repo, but I think that was mostly because the prompts/questions were using h1/h2 headers instead of bold like you are, so I'm not going to push to have the ``` note to be a comment, just wanted to let it be known since I was pleasantly surprised when I saw that.

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
+### Contributing to ZFS on Linux
+
+<p align="center"><img src="http://zfsonlinux.org/images/zfs-linux.png"/></p>
+
+*First of all, thank you for taking the time to contribute!*
+
+The following is a set of guidelines for contributing to ZFS on Linux. 
+Using them you can help us to make a better work.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Perhaps something like `By using the following guidelines, you can help us make ZFS on Linux even better.`?

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+[How Can I Contribute?](#how-can-i-contribute)
+
+  * [Reporting Bugs](#reporting-bugs)
+  * [Suggesting Enhancements](#suggesting-enhancements)
+  * [Pull Requests](#pull-requests)
+  * [Testing](#testing)
+  
+[Styleguides](#styleguides)
+ 
+  * [Coding conventions](#coding-conventions)
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should be `Coding Conventions`

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+### Debugging ZFS
+
+`zdb` utility is the main tool to debug ZFS. 
+
+Additionally, see 
+[build instruction](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS)
+, there is `--enable-debug` flag, which can be set to enable all ASSERTs and additional correctness tests.
+
+### Where can I ask for help?
+The best place to ask is our [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should be `The mailing list is the best place to ask for help.`

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+Please *note*, our issue tracker is **only for bugs**, if you have thoughts about any
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should be two sentences. `Please *note*, our issue tracker is **only for bugs**. If you have any comments or suggestions regarding any enhancements, please use our mailing list.`

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+Please *note*, our issue tracker is **only for bugs**, if you have thoughts about any 
+enhancements - first tell about them via 
+[mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
+
+## How Can I Contribute?
+
+### Reporting Bugs
+
+You can open a new issue and search existing issues using the public
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: I would encourage people to search first. This will help cut down on duplicate issues. Maybe something like `If you run into an issue, please search our issue tracker to ensure the issue hasn't been reported before. If you are seeing a new issue, please create a new issue.` Thoughts?

File: .github/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+[issue tracker](https://github.com/zfsonlinux/zfs/issues). 
+
+Please make sure that what you're actually seeing is a bug and **not a support issue**. 
+
+If in doubt, please ask on the [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists)
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Something like `Please contact use via the mailing list if you aren't certain that you are experiencing a bug.` Thoughts?

File: .github/CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+You can open a new issue and search existing issues using the public
+[issue tracker](https://github.com/zfsonlinux/zfs/issues). 
+
+Please make sure that what you're actually seeing is a bug and **not a support issue**. 
+
+If in doubt, please ask on the [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) 
+first, and if you're then asked to file an issue, do so.
+
+####When opening a new issue include this information at the top of the issue:
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `When opening a new issue, please include the following information at the top of the issue:`?

File: .github/CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+* What distribution you're using and the version.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Maybe `What distribution (with version) you are using.`?

File: .github/CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+* What spl/zfs packages you're using and the version.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Maybe `The spl and zfs versions you are using.`?

File: .github/CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+* Describe the problem you're observing.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `Describe the issue you are experiencing`?

File: .github/CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+* Describe how to reproduce the problem.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Change `problem` to `issue`

File: .github/CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+* Including any warning/errors/backtraces from the system logs.
+
+When a new issue is opened it's not uncommon for a developer to request additional
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Need a comma and slight rewording `When a new issue is opened, it is not uncommon for developers to request additional information.`

File: .github/CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+* Including any warning/errors/backtraces from the system logs.
+
+When a new issue is opened it's not uncommon for a developer to request additional 
+information about the problem. 
+
+In general, the more detail you share about a problem the quicker a developer 
+can resolve it. For example, providing a simple test case is always 
+exceptionally helpful.  
+
+Be prepared to work with the developer looking in to your bug in order to get
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `Be prepared to work with the developer investigating your issue. Your assistance is crucial in providing a quick solution.` Thoughts?

File: .github/CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+**Include any warning/errors/backtraces from the system logs**
+
+Wrap log text with ```
+
+[any log, see https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#reporting-bugs]
+```
+
+### Suggesting Enhancements
+
+For now [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) is the best place
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `Currently, the mailing list is the best place...`?

File: .github/CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+to discuss any enhancements.
+
+ZFS on Linux is stable, and the main priorities for our team are cleaning
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `We attempt to keep ZFS on Linux in a stable state. The current priorities are cleaning up the code, fixing any bugs, and improving performance.`? By saying `main priorities` it sounds like we aren't adding new features often. Thoughts?

File: .github/CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+For now [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) is the best place
+to discuss any enhancements.
+
+ZFS on Linux is stable, and the main priorities for our team are cleaning 
+the code, fixing any bugs and improving performance.
+
+
+### Pull Requests
+
+* One commit inside pull request recommended. Make
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `Please attempt to limit pull requests to a single commit. Be sure to squash multiple commits by performing a rebase.` Not sure you want to restrict a pull request to just a single large commit though. Thoughts?

File: .github/CONTRIBUTING.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+* If your pull request improves performance, please include some benchmarks.
+
+### Testing
+
+We appreciate any help. If you are ready to use bleeding edge code, you can 
+help us to improve ZFS on Linux further by issuing any bugs and performance problems.
+
+Additionally, you can use our [Test Suite](https://github.com/zfsonlinux/zfs/tree/master/tests).
+
+## Styleguides
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `Style Guides`

File: .github/CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+
+### Coding conventions
+
+We use [C  Style  and  Coding  Standards  for  SunOS](http://www.cis.upenn.edu/%7Elee/06cse480/data/cstyle.ms.pdf)
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `We currently use the C Style and Coding Standards for SunOS as our coding convention.`

File: .github/CONTRIBUTING.md [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+* One commit inside pull request recommended. Make 
+[rebase](https://git-scm.com/docs/git-rebase) (squash).
+* Several small pull requests are better than one huge. If you add pull request with several 
+independent improvements, make sequential parts, small pull requests are easier to review and 
+approve, it'll speed up merge process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us 
+with comments and simple code.
+* If your pull request improves performance, please include some benchmarks.
+
+### Testing
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: @behlendorf, should we have a small blurb about the ZFS buildbot here?

File: .github/CONTRIBUTING.md [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+### Get ZFS
+
+You can build fresh zfs packages using 
+[this instruction](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS), 
+or use stable packages from 
+[your distribution repository](https://github.com/zfsonlinux/zfs/wiki/Getting-Started).
+
+### Debugging ZFS
+
+`zdb` utility is the main tool to debug ZFS.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Need `The` out in front of this sentence. Instead of `main` maybe use `primary`?

File: .github/CONTRIBUTING.md [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+  * [Where can I ask for help?](#where-can-I-ask-for-help)
+  
+[How Can I Contribute?](#how-can-i-contribute)
+
+  * [Reporting Bugs](#reporting-bugs)
+  * [Suggesting Enhancements](#suggesting-enhancements)
+  * [Pull Requests](#pull-requests)
+  * [Testing](#testing)
+  
+[Styleguides](#styleguides)
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should this be `Style Guides`?

File: .github/CONTRIBUTING.md [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+####Issue template
+
+```
+**Which distribution are you using (version, kernel)**
+
+[Distribution, version, kernel version, architecture]
+
+**Which versions of spl/zfs packages are you using**
+
+[zfs ver, spl ver]
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Spell out `version`

File: .github/CONTRIBUTING.md [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+independent improvements, make sequential parts, small pull requests are easier to review and 
+approve, it'll speed up merge process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us 
+with comments and simple code.
+* If your pull request improves performance, please include some benchmarks.
+
+### Testing
+
+We appreciate any help. If you are ready to use bleeding edge code, you can 
+help us to improve ZFS on Linux further by issuing any bugs and performance problems.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `We appreciate any help. If you are running the latest code, you can help us improve ZFS on Linux further by reporting general and performance issues you encounter.` Thoughts?

**Comment by dinatale2** [comment]
@gmelikov Thanks for submitting this! I think this is a very good start. I also would be happy to help you iterate on this to perfect it.

File: .github/CONTRIBUTING.md (hunk 4 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 24 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed

File: .github/CONTRIBUTING.md (hunk 5 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed

File: .github/CONTRIBUTING.md (hunk 23 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Thanks, fixed

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 7 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

**Comment by kpande** [comment]
this is fantastic, now we just need users to read it :)

File: .github/CONTRIBUTING.md (hunk 9 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Added it and moved to top of subsection.

File: .github/CONTRIBUTING.md (hunk 10 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Thanks, fixed.

File: .github/CONTRIBUTING.md (hunk 11 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 12 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Added as ` The spl and zfs versions you are using, installation method (repository or manual compilation).`

File: .github/CONTRIBUTING.md (hunk 15 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 16 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 25 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

File: .github/CONTRIBUTING.md (hunk 17 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

**Comment by kpande** [comment]
@gmelikov please stop commenting "fixed" on each request change, just push the changes. you're flooding email and IRC with unnecessary comments.

File: .github/CONTRIBUTING.md (hunk 18 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: I tried to tell about current main goals. I don't want to create a situation then people are striking us by feature requests and waiting that core developers will prioritize them.

I agree, this part may be changed. I'll be happy to any thoughts.

File: .github/CONTRIBUTING.md (hunk 19 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Of course not, that's why only recommended it and added `Several small pull requests are better than one huge. If you add pull request with several 
independent improvements, make sequential parts, small pull requests are easier to review and 
approve, it'll speed up merge process.`

Added your variant. Opened to any thoughts.

File: .github/CONTRIBUTING.md (hunk 21 above)
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: Fixed.

Comment (kpande) [inline .github/CONTRIBUTING.md]: @gmelikov dear god, please stop flooding with these "Fixed." comments.

**Comment by gmelikov** [comment]
@dinatale2 thanks for help1

@kpande sorry, saw it after finished 'fix'ing.

I'll commit changes in several minutes.

File: .github/CONTRIBUTING.md [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+
+  * [ZFS on Linux wiki](https://github.com/zfsonlinux/zfs/wiki)
+  * [OpenZFS Documentation](http://open-zfs.org/wiki/Developer_resources)
+
+
+## What should I know before I get started?
+
+### Get ZFS
+
+You can build fresh zfs packages using
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: `You can build zfs packages by following these instructions or install stable packages from your distribution's repository.`

File: .github/CONTRIBUTING.md [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+You can build fresh zfs packages using 
+[this instruction](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS), 
+or use stable packages from 
+[your distribution repository](https://github.com/zfsonlinux/zfs/wiki/Getting-Started).
+
+### Debugging ZFS
+
+The `zdb` utility is the primary tool to debug ZFS. 
+
+Additionally, see
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should be `Additionally, using the `--enable-debug` flag when building zfs will enable all ASSERTs and additional correctness tests. Refer to the [build instructions](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS#configure-and-build) for more information.`

**Comment by dinatale2** [comment]
It may also be worth mentioning somewhere in the guidelines that test cases should be provided when appropriate.

File: .github/CONTRIBUTING.md [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+
+1. [First Step]
+2. [Second Step]
+3. [Other Steps...]
+
+**Include any warning/errors/backtraces from the system logs**
+
+Wrap log text with ```
+
+[any log, see https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#reporting-bugs]
```
Comment (gmelikov) [inline .github/CONTRIBUTING.md]: I can add instructions to the issue template, but best way to do is as comments (so URL won't render). Or each user has to delete it from template each time.

Issue template section is from [guidance](https://mozillascience.github.io/working-open-workshop/contributing/), which @behlendorf proposed, and it is a good place to add more information and comments to it.

**Comment by gmelikov** [comment]
@dinatale2 first of all, thank you!

Added your changes.

> Should move the "any log" line out and put it under the ``` wrapped text. The URL is not being rendered. Are you able to add instruction text to the issue template? If so, the issue template section may not be necessary.

I'll think what can i do with issue template to improve it.

> It may also be worth mentioning somewhere in the guidelines that test cases should be provided when appropriate

Added  it to pull request chapter.

File: .github/CONTRIBUTING.md [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+
+### Pull Requests
+
+* Please attempt to limit pull requests to a single commit. Be sure to squash multiple commits by performing a [rebase](https://git-scm.com/docs/git-rebase) (squash).
+* Several small pull requests are better than one huge. If you add pull request with several 
+independent improvements, make sequential parts, small pull requests are easier to review and 
+approve, it'll speed up merge process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us 
+with comments and simple code.
+* If your pull request improves performance, please include some benchmarks.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Spoke to @behlendorf offline. Don't need to mention the PERF builder.

**Comment by gmelikov** [comment]
@dinatale2 i added comments to ISSUE_TEMPLATE.md. If you want, i can remove issue template from CONTRIBUTING.md.

Further, i reworked README.markdown, example [here](https://github.com/gmelikov/zfs). Added contribution information with link to it and useful links. And logo.

I'm ready to make PULL_REQUEST_TEMPLATE.md too (example - [node.js](https://github.com/nodejs/node/blob/master/.github/PULL_REQUEST_TEMPLATE.md)), but i need some main points for it (tests requirements, or any other).

**Comment by dinatale2** [comment]
@gmelikov For the pull request template, perhaps [this](https://github.com/dinatale2/zfs/blob/pull_req_template/PULL_REQUEST_TEMPLATE.md) may be a good starting point.

**Comment by gmelikov** [comment]
@dinatale2 thank you, the template is great, added it and rebased commits.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+<!--- Include details of your testing environment, and the tests you ran to -->
+<!--- see how your change affects other areas of the code, etc. -->
+<!--- If your change is a performance enhancement, please provide benchmarks here. -->
+
+## Types of changes
+<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to change)
+
```
Comment (dinatale2) [inline .github/PULL_REQUEST_TEMPLATE.md]: If you feel it would be useful, I don't see why not. @behlendorf what are your thoughts?

**Comment by dinatale2** [comment]
@gmelikov Go ahead and remove the issue template from the contributing guidelines. The comments in the template should be sufficient and you do mention what information we expect when reporting an issue. Having the issue template in the contributing guidelines seems redundant at this point. Thanks!

**Comment by gmelikov** [comment]
@dinatale2 removed it. I redesigned system information in [issue template](https://github.com/gmelikov/zfs/issues/new), now it's a table (i think it'll be easier to read)

File: .github/ISSUE_TEMPLATE.md [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+This issue tracker is for bugs and issues only.
+If you require more general support please use 
+our mailing list. https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists
+
+Please fill in as much of the template below as you're able.
+-->
+
+### System information
+<!--  add version after "|" character -->
+type                                | version/name
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: I don't think the headings are necessary to be honest.

File: .github/ISSUE_TEMPLATE.md [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+  ---                                  |     --- 
+Distribution name       | 
+Distribution version    | 
+Linux kernel                 | 
+Architecture                 | 
+ZFS                                |
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: Make this `ZFS Version` to be clear maybe?

File: .github/ISSUE_TEMPLATE.md [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+### System information
+<!--  add version after "|" character -->
+type                                | version/name
+  ---                                  |     --- 
+Distribution name       | 
+Distribution version    | 
+Linux kernel                 | 
+Architecture                 | 
+ZFS                                | 
+SPL                                |
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: `SPL Version` for clarity?

File: .github/CONTRIBUTING.md [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+
+
+### Pull Requests
+
+* Test cases should be provided when appropriate.
+* Please attempt to limit pull requests to a single commit. Be sure to squash multiple commits by performing a [rebase](https://git-scm.com/docs/git-rebase) (squash).
+* For large pull requests, please consider breaking down your changes into independent
+improvements or sequential parts (if possible). Small pull requests are easier to review and
+approve which speeds up the merging process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: This point might be better as `Try to keep pull requests simple. Simple code with comments is much easier to review and approve.

File: .github/ISSUE_TEMPLATE.md [hunk 36]
```diff
@@ -0,0 +1,37 @@
+<!--
+Thank you for reporting an issue.
+
+This issue tracker is for bugs and issues only.
+If you require more general support please use 
+our mailing list. https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists
+
+Please fill in as much of the template below as you're able.
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: `Please fill in as much of the template as possible.`?

File: .github/ISSUE_TEMPLATE.md [hunk 37]
```diff
@@ -0,0 +1,37 @@
+<!--
+Thank you for reporting an issue.
+
+This issue tracker is for bugs and issues only.
+If you require more general support please use
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: `If you need general support..`?

File: .github/ISSUE_TEMPLATE.md [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+Please fill in as much of the template below as you're able.
+-->
+
+### System information
+<!--  add version after "|" character -->
+type                                | version/name
+  ---                                  |     --- 
+Distribution name       | 
+Distribution version    | 
+Linux kernel                 |
```
Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: You may want to capitalize Name, Version, and Kernel above.

File: .github/ISSUE_TEMPLATE.md (hunk 32 above)
Comment (gmelikov) [inline .github/ISSUE_TEMPLATE.md]: I agree, but, unfourtunately, github won't render table without header.

Comment (dinatale2) [inline .github/ISSUE_TEMPLATE.md]: Bah. Ok, then just capitalize Type, Name, and Version.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 31 above)
Comment (behlendorf) [inline .github/PULL_REQUEST_TEMPLATE.md]: @gmelikov good idea, we have enough code cleanup changes where I think that makes sense.  We also see a lot of performance focused changes so an entry for that would make sense.

**Comment by gmelikov** [comment]
@dinatale2 i added your changes.

**Comment by gmelikov** [comment]
@behlendorf thank you for additions! Added your changes and appended pull request template by code cleanup and performance improvement types of changes. I moved breaking change type to bottom, it's more noticeable.

File: .github/CONTRIBUTING.md [hunk 39]
```diff
@@ -0,0 +1,152 @@
...
+["Feature" label](https://github.com/zfsonlinux/zfs/issues?q=is%3Aopen+is%3Aissue+label%3AFeature).
+If you have an idea for a feature first check this list.  
+If your idea already appears then add a +1 to the top most comment, this helps us gauge interest in that feature.
+
+Otherwise, open a new issue and describe your proposed feature.  Why is this feature needed?  What problem does it solve?
+
+
+### Pull Requests
+
+* All pull requests must be based of the current master branch and apply without conflicts.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Should be `based on the current`

File: .github/CONTRIBUTING.md [hunk 40]
```diff
@@ -0,0 +1,152 @@
...
+  * Whether it is running under a VMM/Hypervisor.
+  * Kernel version.
+  * Values of the spl/zfs module parameters.
+* Stack traces which may be logged to `dmesg`.
+
+
+### Suggesting Enhancements
+
+ZFS on Linux is a widely deployed production filesystem which is under active development.
+The team's primary focus is on fixing known issues, improving performance and adding compelling new features.
```
Comment (dinatale2) [inline .github/CONTRIBUTING.md]: Need a comma after performance. Should be `..known issues, improving performance, and...`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 41]
```diff
@@ -0,0 +1,32 @@
...
+<!--- Please describe in detail how you tested your changes. -->
+<!--- Include details of your testing environment, and the tests you ran to -->
+<!--- see how your change affects other areas of the code, etc. -->
+<!--- If your change is a performance enhancement, please provide benchmarks here. -->
+
+### Types of changes
+<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Performance enhancement (non-breaking change which improve efficiency)
```
Comment (dinatale2) [inline .github/PULL_REQUEST_TEMPLATE.md]: Should be `improves efficiency`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 42]
```diff
@@ -0,0 +1,32 @@
...
+
+### Checklist:
+<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
+<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
+- [ ] My code follows the ZoL code style requirements.
+- [ ] I have updated the documentation accordingly.
+- [ ] I have read the **CONTRIBUTING** document.
+- [ ] I have added tests to cover my changes.
+- [ ] All new and existing tests passed.
+- [ ] Change has been approved by a ZoL member.
```
Comment (dinatale2) [inline .github/PULL_REQUEST_TEMPLATE.md]: Ditch the ZoL acronym to be consistent. Say `ZFS on Linux member`

File: .github/CONTRIBUTING.md [hunk 43]
```diff
@@ -0,0 +1,152 @@
...
+[these instructions](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS), 
+or install stable packages from 
+[your distribution's repository](https://github.com/zfsonlinux/zfs/wiki/Getting-Started).
+
+
+### Debug ZFS
+
+A variety of methods and tools are available to aid ZFS developers.  It's strongly recommended
+that when developing a patch the `--enable-debug` configure option should be set.
+This will enable additional correctness checks and all the ASSERTs to help quickly catch potential issues.
```
Comment (behlendorf) [inline .github/CONTRIBUTING.md]: Since this markdown file will live in the repository please make sure all the lines are wrapped at 80 characters so it'll be readable as plain text.

**Comment by gmelikov** [comment]
@dinatale2 thanks, fixed.
@behlendorf wrapped it.

I removed this text from 'ask fo help' section:
```
Please *note*, our issue tracker is **only for bugs**. If you have any comments or suggestions regarding any enhancements, please use our [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
```

**Comment by behlendorf** [comment]
Merged!  Thanks everybody!
